### ISSUE TITLE
Use "sorted sets" in redis for true priority support

### ISSUE THREAD

**Issue by ashb** [celery/kombu#1178]
I was looking at task priority support from Celery using redis as a broker and after a bit of digging I see why it doesn't work how I expected it at first.

The `priority_steps` setting in broker_transport_options effectively configures a number of "sub"lists/queues, and Kombu then reads from each queue in priority order. This leads to sub-optimal priority, espeically when the possible task priorities aren't know in advance. (I work on Apache Airflow, and the task priorities are dynamic, based on the cumulative priorities of downstream tasks by default, so task with more things depending on it has higher priority.)

In looking at the Redis docs I noticed it has a built in concept of Sorted Sets - https://redis.io/commands#sorted_set

I think it would give a nicer behaviour when using task priorities to use a single sorted set instead of multiple lists. It would:

- Require no tweaking/advanced knowledge of possible values of task priority
- Support (essentially) arbitrary priorities (64bit floating-point double)

Possible downsides:
- More load on Redis to ask for the next values.

Does this feature make sense? If so I (may) at some point tackle it. I just wanted to record the issue and get some feedback.

**Comment by auvipy** [comment]
I'm no expert on Redis, can you dig more on Redis related codes?

**Comment by matusvalo** [comment]
@ashb I think the best approach to the same as is done in RabbitMQ broker implementation. To separate two concepts:
1. normal queues implemented by single simple list with O(1) complexity
2. priority queues implemented as you proposed.

This will not introduce performance degradation in normal queues where is no need for sort functionality and introduce better ordering when needed. Any other ideas?

**Comment by ashb** [comment]
Thanks, I'll take a look at how Rabbitmq does this.

Does this need a transport config change at the front end to enable it, or is it transparently used when enqueing anything with a priority?

**Comment by matusvalo** [comment]
@ashb I prefer to do it as it is already in kombu - see:

https://github.com/celery/kombu/blob/0d3b1e254f9178828f62b7b84f0307882e28e2a0/t/integration/common.py#L206-L244

I have one important question regarding ordered set in redis. Does it keeps order of messages with the *same* priority? E.g. When messages A, B, C are published by producer in this order with the same priority, will be also fetched by Consumer in the same order - A, B, C?

**Comment by ashb** [comment]
Not sure, I will check

**Comment by ashb** [comment]
https://redis.io/commands/zadd#elements-with-the-same-score

> While the same element can't be repeated in a sorted set since every element is unique, it is possible to add multiple different elements having the same score. When multiple elements have the same score, they are ordered lexicographically (they are still ordered by score as a first key, however, locally, all the elements with the same score are relatively ordered lexicographically).
>
> The lexicographic ordering used is binary, it compares strings as array of bytes.

So no, not unless we made the insertion time the first thing in the value so it sorts correctly for us

**Comment by matusvalo** [comment]
@ashb  unfortunatelly, in this case sorted sets cannot be used. But I have a proposal. Not sure if it make sense since I am pretty new in redis transport.

I propose the similar way how to work as it is in AMQP transport by putting `max_priority` to `Queue`

```python
        test_queue = kombu.Queue(
            'priority_test', routing_key='priority_test', max_priority=10
        )
```

Instead of using `4` lists each for each step in `PRIORITY_STEPS = [0, 3, 6, 9]` when `max_priority=n` is set as parameter,  I propose to simply create `n` queues.

Hence, for each priority we will have single list. Regarding perfromance, kombu is using BRPOP for reading data from multiple queues:

https://github.com/celery/kombu/blob/5101db177c1b2cee0febcc64ad0eb49b39c775e8/kombu/transport/redis.py#L732

Hence:
1. Code will be more readable
2. The behavior will less obscure and more close to AMQP - data with higher priority will be read first
3. Data with same priority will be sorted as it is now.

The only open point is what is biggest number of priorities we can support this way - or more explicitely - from how many queues can BRPOP read at once?

**Comment by ashb** [comment]
@matusvalo That would be better yes, but still far from ideal for my usecase.

So looking at how the values we get back from redis are handled I see:

https://github.com/celery/kombu/blob/5101db177c1b2cee0febcc64ad0eb49b39c775e8/kombu/transport/redis.py#L745-L749

My (very rough) plan was something like this on the sending side:

```python
        with self.conn_or_acquire() as client:
            redis_message = "{}{}{}".format(str(time.time()), self.sep, dumps(message))
            client.zadd(queue, pri, redis_message)
```

And change it something like this on receiving:

```python
            if dest__item:
                dest, item = dest__item
                dest = bytes_to_str(dest).rsplit(self.sep, 1)[0]
                self._queue_cycle.rotate(dest)
                item = bytes_to_str(item).rsplit(self.sep, 1)[1]
                self.connection._deliver(loads(item), dest)
```

I'm sure I've missed something, but wouldn't that work in general principle?

**Comment by matusvalo** [comment]
@ashb I am not sure if I understood you correctly. As long as I understand you are proposing to solve the problem with sorting of messages using timestamp on the client side. It means that client fetches data from sorted set and orders them using timestamp. Correct me if I am wrong. In this case I see the problem that consumer needs to fetch all data to be able to sort it. E.g. what will happen if you have 1M messages in queue (sorted set) with the same priority? Consumer needs to fetch them all and sort it accordingly?

**Comment by ashb** [comment]
No, I'm not proposing a client side fix, all sorting still handled by redis.

So redis returns objects from a sorted set, it first sorts by score, and then orders by the value itself.

For example, if we just put the messages in a Sorted Set directly, the return value for messages with the same priority depends on how the "values" sort.

So for example:
```
In [4]: r = redis.Redis(host='localhost', port=6379, db=0)                                                                                                                         

In [6]: r.zadd('my_set', {'VALUE': 10})                                                                                                                                            
Out[6]: 1

In [7]: r.zadd('my_set', {'ABC': 10})
Out[7]: 1
```

Here, even tho `ABC` is added later it would be returned first, since it sorts lower than `VALUE`. Not what we want.

```
In [9]: r.zpopmin('my_set')                                                                                                                                                        
Out[9]: [(b'ABC', 10.0)]

In [10]: r.zpopmin('my_set')                                                                                                                                                       
Out[10]: [(b'VALUE', 10.0)]
```

(Assume for argument here that `VALUE` and `ABC` were how Celery encoded messages. It's clearly not, but it's an easier example to use)

To work around that is why I add the "sorting/date" prefix.

Let's try that

```
In [14]: r.zadd('my_set', {"{0:019.7f}|{1}".format(time.time(), 'VALUE'): 10})                                                                                                     
Out[14]: 1

In [15]: r.zadd('my_set', {"{0:019.7f}|{1}".format(time.time(), 'ABC'): 10})                                                                                                       
Out[15]: 1
```

And lets see what order that returns values in:

```
In [14]: r.zadd('my_set', {"{0:019.7f}|{1}".format(time.time(), 'VALUE'): 10})                                                                                                     
Out[14]: 1

In [15]: r.zadd('my_set', {"{0:019.7f}|{1}".format(time.time(), 'ABC'): 10})                                                                                                       
Out[15]: 1
```

Much better!.

And just for completness, lets try with a different score/priority:

```
In [18]: r.zadd('my_set', {"{0:019.7f}|{1}".format(time.time(), 'VALUE'): 10})                                                                                                     
Out[18]: 1

In [19]: r.zadd('my_set', {"{0:019.7f}|{1}".format(time.time(), 'ABC'): 5})                                                                                                        
Out[19]: 1

In [20]: r.zpopmin('my_set')                                                                                                                                                       
Out[20]: [(b'01589930360.3527460|ABC', 5.0)]

In [21]: r.zpopmin('my_set')                                                                                                                                                       
Out[21]: [(b'01589930351.2135429|VALUE', 10.0)]
```

So all the sorting is handled by the redis server/cluster, the only operation performed by the client is to add or strip this prefix.

Does that help explain it @matusvalo?
